lms/djangoapps/gating/api.py

```diff
@@ -61,7 +61,7 @@ def _get_subsection_percentage(subsection_grade):
     """
     Returns the percentage value of the given subsection_grade.
     """
-    return _calculate_ratio(subsection_grade.graded_total.earned, subsection_grade.graded_total.possible) * 100.0
+    return int("{0:.0f}".format(_calculate_ratio(subsection_grade.graded_total.earned, subsection_grade.graded_total.possible) * 100.0))
```

Comment:
Why is the .format here? We shouldn't need to convert a float to a string just to convert it to an integer.  Are we rounding? Or flooring? There are functions for those operations.

Reply:
For exact halfway cases (e.g. 2.50); **"{0:.0f}".format()** rounds the number to the nearest even number, and python 2.7's **round()** rounds the number to the next whole number. **"{0:.0f}".format()** would round 2.50 to 2 and 1.50 to 2. On the other hand **round()** would round 2.50 to 3 and 1.50 to 2.

In the two other places, subsection percentage is displayed, **"{0:.0%}".format()** is doing the rounding, for sake of consistency, we use the same method of rounding. And since we need an int instead of a string, we have to convert.

Reply:
You say there are other places where "{0:.0%}" is used, but I can't find them.  If there are multiple places that need specific rounding, we should definitely write and use a tested helper function to get it right.

I'm also confused why we want 2.5 to round to 2, but 3.5 to round to 4?  I understand the need for that in some domains, but why for grading?  Is that what people will expect?

Reply:
line number 61 of `lms/templates/courseware/progress.html`:
`percentageString = "{0:.0%}".format( float(earned)/total) if earned > 0 and total > 0 else ""`
rounded percentage is used in progress page in the graph as well as score listings below the graph.

We can use either, we just have to make sure same method of rounding is used in all 3 places (progress graph, score list below the graph and in subsection gating), generally people expect rounding to go to next integer (2.5 to 3) because that's what is taught in schools, and is more generous towards the student while rounding to even number wouldn't be as generous to the student, sometimes it would favor the student and sometimes it would not.

Reply:
@edx/educator-devs Please look into this rounding issue, and address it in a definitive way that will prevent confusion among future developers.  I suspect that the original author of "{0:.0%}".format() didn't use it because of the even-up odd-down rounding behavior.

Reply:
@AmmarNaqvi @nasthagiri 
We've replaced the call to _calcuateRatio() with getting the percentage directly from subsection_grade.percent_graded and moved it to another file in this PR:
https://github.com/edx/edx-platform/pull/16876/files#diff-bca4c6ed3cdc85414872cf10fee0fa41R408
That PR probably won't land for some time still so if we determine this is the proper solution I can incorporate it in my PR after this lands.

Reply:
@AmmarNaqvi A couple of things:
1. How we round up a subsection percentage seems like business logic that should be moved down into the Grades django app layer - rather than duplicated in multiple presentation layers.  For example, grade reports would also need to be updated to have this rounding logic.
2. Given that, can we move this logic down into [NonZeroSubsection.percent_graded](https://github.com/edx/edx-platform/blob/b6030a4b136762ed64f51b1e37f0d11ecfd2335f/lms/djangoapps/grades/subsection_grade.py#L143)?
3. Update the current UI code (you linked above) to get the value as computed by `NonZeroSubsection`?
4. Update this code to use `NonZeroSubsection.percent_graded` instead of  `_calcuateRatio` (which @bill-filler is planning to do in his PR as well).

Reply:
Regarding @nedbat's point, why does 2.5 round down to 2?  Is that a bug in python's formatting code?

Following https://gist.github.com/jackiekazil/6201722, can we use the `round` function instead within `NonZeroSubsection.percent_graded`?

Reply:
@nasthagiri It is not actually a bug, since python 3.x round() method follows the same method for rounding. 
https://docs.python.org/3/library/functions.html#round
https://docs.python.org/2/library/functions.html#round

Reply:
I agree, using `NonZeroSubsection.percent_graded` seems like a better idea. UI code can be updated accordingly as well. But that isn't implemented till @bill-filler 's PR is merged.

Reply:
Yes, copying that into this PR sounds good in order to unblock you.  @bill-filler has offered to resolve any merge conflicts.

Reply:
Also, I suggest making sure Product (@sstack22) is aware that we are changing the code to Round to the Nearest Even, per https://en.wikipedia.org/wiki/IEEE_754#Rounding_rules.

Reply:
Have you thought through if there are any backward compatibility issues with this change?